### DISCUSSION TITLE
cannot Copy content from one editor to another.

### DISCUSSION THREAD

**Discussion by we-wake-act21** [eclipse-glsp/glsp#823] [Q&A]
I want to copy content from one open editor to another.
Is there any implementation or reference regarding this?

**Comment by planger** [comment]
Hi,

alas we only have a default implementation to copy&paste within one diagram. However, this can be extended to also work across diagrams or even applications. An implementation will likely depend on your specific diagram use case, as you'll have to specify the text that is put into the clipboard and be able to identify what needs to be done when a certain text from the clipboard is pasted into your diagram.

Parts of the copy&paste implementation also depend on the platform integration (Theia, VS Code, Standalone, Eclipse), as those handle interaction with the clipboard differently. So it would be helpful to know which integration you are targeting and which server you are using. At the moment clipboard handling is only supported on the Java GLSP server framework and in Theia, Eclipse and VS Code.

I'll give some pointers for Java and Theia:

- If a user invokes the copy, the client will send a [`RequestClipboardDataAction`](https://github.com/eclipse-glsp/glsp/blob/master/PROTOCOL.md#2171-requestclipboarddataaction). This is initiated in the default implementation of the [`ServerCopyPasteHandler`](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/copy-paste/copy-paste-handler.ts#L102).
- Your server then needs to have an `ActionHandler` for the `RequestClipboardDataAction` registered and replies with a [`SetClipboardAction`](https://github.com/eclipse-glsp/glsp/blob/master/PROTOCOL.md#2172-setclipboarddataaction) that contains the data to be put into the client clipboard.
- The client hands over the data of this `SetClipboardAction` to its implementation of the `ClipboardService`. Again see [`ServerCopyPasteHandler`](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/copy-paste/copy-paste-handler.ts#L102).
- By default, GLSP just binds a [`LocalClipboardService`](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/copy-paste/copy-paste-handler.ts#L53) implementation to the `ClipboardService` interface, which just puts the data into memory, but not actually into the clipboard. Instead, it puts an id into the clipoard that refers to the clipboard data stored in the memory. This is just because a browser only can store text into the clipboard and for other use cases we wanted to support arbitrary data types (such as application/json) sent from the server. Thus, in order to support cross-application copy & paste, you'd need to rebind a specific implementation of the `ClipboardService` and potentially also `ServerCopyPasteHandler` that actually writes the contents of the clipboard data sent from the server into the users' clipboard and reads it on paste.

On paste:
- By default, the `ServerCopyPasteHandler` reads the clipboard data on paste, but assumes it is an ID refering to the clipboard data stored in memory (see last item in the list above). For that ID it obtains the stored clipboard data and sends it to the server in a `PasteOperation`.
- The server must have an `OperationHandler` registered that processes this `PasteOperation`, ie reading the provided clipboard data and deciding what to add into the diagram.
